File: contrib/python/potatoe.py
#!/usr/bin/env python
import random
import sys
import textwrap
import argparse


def terminal_size():
    w = 100
    h = 100
    try:
        import fcntl
        import termios
        import struct
        h, w, _, _ = struct.unpack('HHHH', fcntl.ioctl(0, termios.TIOCGWINSZ,
                                   struct.pack('HHHH', 0, 0, 0, 0)))
    except:
        pass
    return w, h


tmpl = r'''
        \        ___--===--___
         \    __=     ___   - \
            _/     o           |
         /==   \     __-- o    |
        |   o   -            _/
         \__    \    -   o //
          -===============-       - dan quayle'''


def show(quote, width):
    out = ""
    wraps = textwrap.wrap(quote, width)
    if len(wraps) == 0:
        wraps.append("...")
    maxwidth = len(wraps[0])
    for line in wraps:
        if len(line) > maxwidth:
            maxwidth = len(line)
    out = out + " " + "_" * (maxwidth+2) + "\n"
    for idx, line in enumerate(wraps):
        if idx == 0:
            out = out + ("< " if len(wraps) == 1 else "/ ")
        elif (idx == (len(wraps) - 1)):
            out = out + "\\ "
        else:
            out = out + "| "
        out = out + line.ljust(maxwidth, " ")
        if idx == 0:
            out = out + (" >" if len(wraps) == 1 else " \\")
        elif (idx == (len(wraps) - 1)):
            out = out + " /"
        else:
            out = out + " |"
        out = out + "\n"
    out = out + " " + "-" * (maxwidth+2)

    out = out + tmpl

    print(out)

# get the quote list


def loadFile(path, quoteL):
    try:
        with open(path) as f:
            content = f.readlines()
            for k in content:
                if len(k) > 4:
                    if k[0] != "#":
                        quoteL.append(k.replace("\n", ""))
    except:
        pass


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("-t", "--text",
                        help="show only the text",
                        action="store_true")

    parser.add_argument("-w", "--width",
                        help="override width",
                        )
    args = parser.parse_args()
    if args.width is not None:
        defaultFill = int(args.width)
    else:
        (w, h) = terminal_size()
        defaultFill = 0.64 * w
    quoteL = []
    for fileName in [
            "./quotes.txt",
            "/usr/lib/potatoe/quotes.txt",
            "/var/lib/potatoe/quotes.txt",
    ]:
        loadFile(fileName, quoteL)
    # ok we loaded the quote list, now filter unique ones
    if len(quoteL) == 0:
        quoteL = ["no quotes"]
    quoteL = list(set(quoteL))
    # now select a random one
    toread = random.randint(0, len(quoteL)-1)
    selected = quoteL[toread]
    if args.text:
        print(selected)
        return
    show(selected, defaultFill)
    return 0


if __name__ == '__main__':
    sys.exit(main())

File: contrib/kotlin/potatoe.kt
#!/usr/bin/env kotlinc -script

import kotlinx.cinterop.*
import platform.posix.*
import kotlin.random.Random
import kotlin.system.exitProcess

@OptIn(kotlinx.cinterop.ExperimentalForeignApi::class)
fun terminalSize(): Pair<Int, Int> {
    var w = 100
    var h = 100

    try {
        memScoped {
            val winsize = alloc<winsize>()
            if (ioctl(STDOUT_FILENO.convert(), TIOCGWINSZ.convert(), winsize.ptr) != -1) {
                h = winsize.ws_row.toInt()
                w = winsize.ws_col.toInt()
            }
        }
    } catch (e: Exception) {
        // Fallback to default values if ioctl fails
    }

    return Pair(w, h)
}

val tmpl = """
        \        ___--===--___
         \    **=     **_   - \
            _/     o           |
         /==   \     __-- o    |
        |   o   -            _/
         \__    \    -   o //
          -===============-       - dan quayle"""

fun show(quote: String, width: Int) {
    var out = ""
    val wraps = wrapText(quote, width)

    if (wraps.isEmpty()) {
        wraps.add("...")
    }

    var maxwidth = wraps[0].length
    for (line in wraps) {
        if (line.length > maxwidth) {
            maxwidth = line.length
        }
    }

    out += " " + "_".repeat(maxwidth + 2) + "\n"

    for ((idx, line) in wraps.withIndex()) {
        out += when {
            idx == 0 -> if (wraps.size == 1) "< " else "/ "
            idx == wraps.size - 1 -> "\\ "
            else -> "| "
        }

        out += line.padEnd(maxwidth, ' ')

        out += when {
            idx == 0 -> if (wraps.size == 1) " >" else " \\"
            idx == wraps.size - 1 -> " /"
            else -> " |"
        }

        out += "\n"
    }

    out += " " + "-".repeat(maxwidth + 2)
    out += tmpl
    println(out)
}

fun wrapText(text: String, width: Int): MutableList<String> {
    val result = mutableListOf<String>()
    var remaining = text

    while (remaining.isNotEmpty()) {
        val splitIndex = if (remaining.length <= width) {
            remaining.length
        } else {
            val lastSpace = remaining.substring(0, width).lastIndexOf(' ')
            if (lastSpace == -1) width else lastSpace
        }

        result.add(remaining.substring(0, splitIndex))
        remaining = if (splitIndex < remaining.length) {
            remaining.substring(splitIndex).trimStart()
        } else {
            ""
        }
    }

    return result
}

@OptIn(kotlinx.cinterop.ExperimentalForeignApi::class)
fun loadFile(path: String, quoteList: MutableList<String>) {
    try {
        val file = fopen(path, "r") ?: return

        try {
            memScoped {
                val bufferLength = 1024
                val buffer = allocArray<ByteVar>(bufferLength)

                while (true) {
                    val line = fgets(buffer, bufferLength, file)?.toKString() ?: break
                    val trimmed = line.replace("\n", "")

                    if (trimmed.length > 4 && trimmed[0] != '#') {
                        quoteList.add(trimmed)
                    }
                }
            }
        } finally {
            fclose(file)
        }
    } catch (e: Exception) {
        // Ignore file read errors
    }
}

fun Main(args: Array<String>): Int {
    var showTextOnly = false
    var customWidth: Int? = null

    // Simple argument parsing
    var i = 0
    while (i < args.size) {
        when (args[i]) {
            "-t", "--text" -> showTextOnly = true
            "-w", "--width" -> {
                if (i + 1 < args.size) {
                    customWidth = args[i + 1].toIntOrNull()
                    i++
                }
            }
        }
        i++
    }

    val defaultFill = if (customWidth != null) {
        customWidth
    } else {
        val (w, _) = terminalSize()
        (0.64 * w).toInt()
    }

    val quoteList = mutableListOf<String>()

    val fileNames = listOf(
        "./quotes.txt",
        "/usr/lib/potatoe/quotes.txt",
        "/var/lib/potatoe/quotes.txt"
    )

    for (fileName in fileNames) {
        loadFile(fileName, quoteList)
    }

    // Handle empty quote list
    if (quoteList.isEmpty()) {
        quoteList.add("no quotes")
    }

    // Remove duplicates
    val uniqueQuotes = quoteList.toSet().toList()

    // Select a random quote
    val toread = Random.nextInt(uniqueQuotes.size)
    val selected = uniqueQuotes[toread]

    if (showTextOnly) {
        println(selected)
        return 0
    }

    show(selected, defaultFill)
    return 0
}

fun main(args: Array<String>) {
    exitProcess(Main(args))
}
